### DISCUSSION TITLE
Prisma update operation failing in validator; using `UncheckedUpdateInput` type instead of `UpdateInput` type

### DISCUSSION THREAD

**Discussion by MikeCarbone** [prisma/prisma#17493] [Q&A] [answered]
Hey! I'm trying to perform an update operation:

```js
  const specialItem = await db.specialItem.update({ where: query, data: formattedUpdate })
```

The value `formattedUpdate` in that statement looks something like this:
```js
{
  item_name: null,
  drink_type: 'Wine',
  price: 800,
  bar: {
    connect: {
      id: '4486bbc3-5ff8-47c8-99cf-9291c57ebe78'
    }
  },
  event: {
    connect: {
      id: undefined
    }
  },
  related_item: {
    connect: {
      id: undefined
    }
  }
}
```

This operation fails in the Prisma validator with the errors:
```
[1] Unknown arg `bar` in data.bar for type SpecialItemUncheckedUpdateInput. Did you mean `id`? Available args: ...
[1] Unknown arg `event` in data.event for type SpecialItemUncheckedUpdateInput. Did you mean `event_id`? Available args: ...
[1] Unknown arg `related_item` in data.related_item for type SpecialItemUncheckedUpdateInput. Did you mean `related_item_id`? Available args: ...
```

However those fields are most definitely in my schema:
```prisma
model SpecialItem {
  id                  String      @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  created_at          DateTime    @default(now()) @db.Timestamptz(3)
  event               Event?      @relation(fields: [event_id], references: [id], onDelete: Cascade)
  event_id            String?     @db.Uuid
  bar                 Bar         @relation(fields: [bar_id], references: [id], onDelete: Cascade)
  bar_id              String      @db.Uuid
  related_item_id     String?     @db.Uuid
  related_item        Item?       @relation(fields: [related_item_id], references: [id])
  is_recurring        Boolean     @default(false)
  is_inactive         Boolean     @default(false)
  ...
}
```

The available args the error suggests comes from the generated Prisma type `ModelNameUncheckedUpdateInput`, however I believe this operation would succeed if it were using the generated Prisma type `ModelNameUpdateInput`. **The _unchecked_ version uses IDs directly, where as the _regular_ version uses the connect object pattern.** I'd prefer using the latter. Why is Prisma expecting me to satisfy the unchecked version of the type? Can I force the validator to use the regular version?

Has anyone run into this before? I don't think it makes sense to break from the connect pattern I use elsewhere.

Using Prisma version 4.9.0.

**Comment by nurul3101** [comment, marked as answer]
Hello @MikeCarbone 👋 

Based on your schema, I tried to reproduce this locally, but the connect pattern works as expected for me.
Can you check if you get the same validator error if you pass a value other than `undefined` in the connect clause?

Considering this schema:
```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["views", "fieldReference"]
}

model SpecialItem {
  id              String   @id @default(dbgenerated("gen_random_uuid()"))
  created_at      DateTime @default(now()) @db.Timestamptz(3)
  event           Event?   @relation(fields: [event_id], references: [id], onDelete: Cascade)
  event_id        String?
  bar             Bar      @relation(fields: [bar_id], references: [id], onDelete: Cascade)
  bar_id          String
  related_item_id String?
  related_item    Item?    @relation(fields: [related_item_id], references: [id])
  is_recurring    Boolean  @default(false)
  is_inactive     Boolean  @default(false)
}

model Event {
  id          String        @id @default(dbgenerated("gen_random_uuid()"))
  name        String
  SpecialItem SpecialItem[]
}

model Bar {
  id          String        @id @default(dbgenerated("gen_random_uuid()"))
  name        String
  SpecialItem SpecialItem[]
}

model Item {
  id          String        @id @default(dbgenerated("gen_random_uuid()"))
  name        String
  SpecialItem SpecialItem[]
}
``` 

This query tries to connect bar, event and related_item models to specialItem model.
```typescript
import { Prisma, PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

// A `main` function so that you can use async/await
async function main() {
  await prisma.specialItem.create({
    data: {
      id: "1",
      bar: {
        create: {
          id: "1",
          name: "bar 1",
        },
      },
      event: {
        create: {
          id: "1",
          name: "event 1",
        },
      },
      related_item: {
        create: {
          id: "1",
          name: "related item 1",
        },
      },
    },
  });

  await prisma.bar.create({
    data: {
      id: "2",
      name: "bar 2",
    },
  });

  await prisma.event.create({
    data: {
      id: "2",
      name: "event 2",
    },
  });

  await prisma.item.create({
    data: {
      id: "2",
      name: "item 2",
    },
  });

  const response = await prisma.specialItem.update({
    where: {
      id: "1",
    },
    data: {
      bar: {
        connect: {
          id: "2",
        },
      },
      event: {
        connect: {
          id: "2",
        },
      },
      related_item: {
        connect: {
          id: "2",
        },
      },
    },
  });

  console.log("response", response);
}

main()
  .catch((e) => {
    throw e;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
```

Here's the response:
<img width="339" alt="Screenshot 2023-01-24 at 6 08 21 PM" src="https://user-images.githubusercontent.com/13049676/214293623-67fc5039-dbbc-4d43-a4c3-037a5181f32a.png">

**Reply by MikeCarbone** [reply]
Thanks for this @nurul3101 ! It was my understanding that `undefined` meant "do nothing", but it seems to mess with the Prisma validator. Removing the undefined values works.

Another point: if you want to unset a relation value, instead of using `connect: { id: null }}`, instead should use `disconnect: true`.

**Reply by nurul3101** [reply]
Yes, you should use disconnect to remove the relation 👍

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
